solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,377 @@
+            fundingOutputIndex
+        );
+
+        uint256 requestedAt = pendingOptimisticMints[depositKey];
+        require(
+            requestedAt != 0,
+            "Optimistic minting not requested or already finalized"
+        );
+        require(
+            /* solhint-disable-next-line not-rely-on-time */
+            block.timestamp - requestedAt > OPTIMISTIC_MINTING_DELAY,
+            "Optimistic minting delay has not passed yet"
+        );
+
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
+        require(deposit.sweptAt == 0, "The deposit is already swept");
+
+        // Bridge, when sweeping, cuts a deposit treasury fee and splits
+        // Bitcoin miner fee for the sweep transaction evenly between the
+        // depositors in the sweep.
+        //
+        // When tokens are optimistically minted, we do not know what the
+        // Bitcoin miner fee for the sweep transaction will look like.
+        // The Bitcoin miner fee is ignored. When sweeping, the miner fee is
+        // subtracted so the optimisticMintingDebt may stay non-zero after the
+        // deposit is swept.
+        //
+        // This imbalance is supposed to be solved by a donation to the Bridge.
+        uint256 amountToMint = deposit.amount - deposit.treasuryFee;
+        _mint(deposit.depositor, amountToMint);
+        optimisticMintingDebt[deposit.depositor] += amountToMint;
+
+        delete pendingOptimisticMints[depositKey];
+
+        emit OptimisticMintingFinalized(
+            msg.sender,
+            deposit.depositor,
+            amountToMint,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    /// @notice Allows a Guardian to cancel optimistic minting request. The
+    ///         following conditions must be met:
+    ///         - The optimistic minting request for the given deposit exists.
+    ///         - The optimistic minting request for the given deposit has not
+    ///           been finalized yet.
+    /// @dev Guardians must validate the following conditions for every deposit
+    ///      for which the optimistic minting was requested:
+    ///      - The deposit happened on Bitcoin side and it has enough
+    ///        confirmations.
+    ///      - The optimistic minting has been requested early enough so that
+    ///        the wallet has enough time to sweep the deposit.
+    ///      - The wallet is an active one and it does perform sweeps or it will
+    ///        perform sweeps once the sweeps are activated.
+    function cancelOptimisticMint(
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex
+    ) external onlyGuardian {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+
+        require(
+            pendingOptimisticMints[depositKey] > 0,
+            "Optimistic minting not requested of already finalized"
```

Comment:
```suggestion
            "Optimistic minting was not requested or it has been finalized"
```

Reply:
This conflicts with [this](https://github.com/keep-network/tbtc-v2/pull/424#discussion_r1047236280) comment. @nkuba @dimpar let's settle on one version. Both options work for me but I'd probably prefer the current (shorter) one from 2823919eae18055e1ca13451511e43d401c9a79e.

Reply:
No biggy, a shorter version is fine by me. There was a typo that Kuba also noticed. But I think we miss `was`, ie `Optimistic minting was not requested or already finalized`